### PR TITLE
[Enhancement] BERT pre-training data generation from sentencepiece vocab

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
## Description ##
- add sentencepiece vocabulary support
- addresses #726
- more documentation
- remove layer norm fp32 casting
- support comma in dataset stream API

### Essentials ###
- [x] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by mli** [comment]
Job PR-743/1 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-743/1/index.html

**Comment by eric-haibin-lin** [comment]
@davisliang FYI

**Comment by mli** [comment]
Job PR-743/3 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-743/3/index.html

**Comment by mli** [comment]
Job PR-743/4 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-743/4/index.html

**Comment by mli** [comment]
Job PR-743/5 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-743/5/index.html

**Comment by mli** [comment]
Job PR-743/6 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-743/6/index.html

**Comment by mli** [comment]
Job PR-743/7 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-743/7/index.html

**Comment by mli** [comment]
Job PR-743/8 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-743/8/index.html

**Comment by mli** [comment]
Job PR-743/9 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-743/9/index.html

**Comment by mli** [comment]
Job PR-743/10 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-743/10/index.html

**Comment by mli** [comment]
Job PR-743/11 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-743/11/index.html

**Comment by paperplanet** [comment]
New create_pretraining_data.py seems slower than the previous version which do multiprocess during the whole reading input text and writing npz file procedure.

File: scripts/bert/run_pretraining_hvd.py [hunk 1]
```diff
@@ -64,7 +65,15 @@
...
+parser.add_argument('--sentencepiece', default=None, type=str,
+                    help='Path to the sentencepiece .model file for both tokenization and vocab.')
+parser.add_argument('--sp_nbest', type=int, default=0,
+                    help='Number of best candidates for sampling subwords with sentencepiece. ')
+parser.add_argument('--sp_alpha', type=float, default=1.0,
+                    help='Inverse temperature for probability rescaling for sentencepiece sampling')
+parser.add_argument('--num_data_workers', type=int, default=8,
+                    help='Number of workers to pre-process data.')
 
 args = parser.parse_args()
```
[the remark this answers is not in the document]
Comment (eric-haibin-lin) [reply scripts/bert/run_pretraining_hvd.py]: Horovod has to be compiled against mxnet. If you have a horovod version without mxnet enabled, it won't work either. I can update the err msg to be more specific

File: scripts/bert/index.rst [hunk 2]
```diff
@@ -207,20 +210,18 @@ The BERT base model produced by gluonnlp pre-training script (`log <https://raw.
...
 
-Then you can install the master version of horovod:
+Then you can install horovod v0.16.2 via the following command:
 
 .. code-block:: console
 
-    $ git clone --recursive https://github.com/uber/horovod horovod;
-    $ cd horovd;
-    $ HOROVOD_GPU_ALLREDUCE=NCCL pip install . --user --no-cache-dir
+    $ HOROVOD_GPU_ALLREDUCE=NCCL pip install horovod --user --no-cache-dir
```
[the remark this answers is not in the document]
Comment (eric-haibin-lin) [reply scripts/bert/index.rst]: sure I can update the instruction

**Comment by mli** [comment]
Job PR-743/12 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-743/12/index.html

**Comment by mli** [comment]
Job PR-743/13 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-743/13/index.html

File: scripts/bert/index.rst [hunk 3]
```diff
@@ -232,8 +233,10 @@ Run pre-training with horovod:
 
 .. code-block:: console
 
+    $ # single node, multi-gpu
```
Comment (szha) [inline scripts/bert/index.rst]: splitting into two code blocks would make it more clear.

File: scripts/bert/index.rst [hunk 4]
```diff
@@ -207,33 +210,34 @@ The BERT base model produced by gluonnlp pre-training script (`log <https://raw.
...
-    $ mpirun -np 16 -H node0:8,node1:8 -mca pml ob1 -mca btl ^openib -mca btl_tcp_if_exclude docker0,lo --map-by ppr:4:socket -x NCCL_MIN_NRINGS=8 -x NCCL_DEBUG=INFO python run_pretraining_hvd.py --batch_size 8192 --accumulate 1 --lr 1e-4 --data "/path/to/generated/samples/train/*.npz" --warmup_ratio 0.01 --num_steps 1000000 --log_interval=250 --ckpt_dir './ckpt' --ckpt_interval 25000 --num_buckets 10 --dtype float16 --use_avg_len --verbose
+
+Run pre-training with horovod on node0 and node1, with 8 GPUs each:
+
+    $ mpirun -np 16 -H node0:8,node1:8 -mca pml ob1 -mca btl ^openib -mca btl_tcp_if_exclude docker0,lo --map-by ppr:4:socket -x NCCL_MIN_NRINGS=8 -x NCCL_DEBUG=WARNING -x HOROVOD_HIERARCHICAL_ALLREDUCE=1 --tag-output python run_pretraining_hvd.py --batch_size 8192 --accumulate 1 --lr 1e-4 --data "/path/to/generated/samples/train/*.npz" --warmup_ratio 0.01 --num_steps 1000000 --log_interval=250 --ckpt_dir './ckpt' --ckpt_interval 25000 --num_buckets 10 --dtype float16 --use_avg_len --verbose
```
Comment (leezu) [inline scripts/bert/index.rst]: This line is raising warnings when generating the documentations: `Inline emphasis start-string without end-string.` (warnings are treated as fatal in our CI setup)

**Comment by mli** [comment]
Job PR-743/15 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-743/15/index.html

File: scripts/bert/index.rst (hunk 4 above)
Comment (eric-haibin-lin) [reply scripts/bert/index.rst]: Thanks. I split them into two separate blocks with `code-block ` annotations

**Pull request merged**